File: python2/kirishima.py
numOfCells = int(raw_input())
goal = numOfCells - 1
cells = raw_input().strip().split(" ")

CURRENT = 0
FINISHED = 1
DEAD = 2

for idx in range(0, int(raw_input())):
    actNum = int(raw_input())
    moveLog = []
    status = CURRENT
    while status == CURRENT:
        if actNum == goal:
            status  = FINISHED
        elif actNum < 1 or goal < actNum:
            status = DEAD
        else:
            moves = int(cells[actNum])
            if moves == 0:
                status = DEAD
            else:
                actNum += moves
                for movedNum in moveLog:
                    if actNum == movedNum:
                        status = DEAD
                        break
                if status != DEAD
                    moveLog.append(actNum)
    print ("Yes" if status == FINISHED else "No")

File: ruby/kirishima.rb
numOfCells = gets.to_i
goal = numOfCells - 1
cells = gets.strip.split(" ")

CURRENT = 0
FINISHED = 1
DEAD = 2

gets.to_i.times {
    actNum = gets.to_i
    moveLog = Array.new

    status = CURRENT

    while status == CURRENT do
        if actNum == goal then
            status = FINISHED
        elsif actNum < 1 || goal < actNum then
            status = DEAD
        else
            moves = cells[actNum].to_i
            if moves == 0 then
                status = DEAD
            else
                actNum += moves
                moveLog.each {|movedNum|
                    if actNum == movedNum then
                        status = DEAD
                        break
                    end
                }
                if status != DEAD then
                    moveLog.push(actNum)
                end
            end
        end
    end
    puts status == FINISHED ? "Yes" : "No"
}
